server/msgpack/server.go

```diff
@@ -0,0 +1,240 @@
+	clockOpts := opts.ClockOptions()
+	instrumentOpts := opts.InstrumentOptions()
+	scope := instrumentOpts.MetricsScope().SubScope("server")
+	queue := packet.NewQueue(opts.PacketQueueSize(), clockOpts, instrumentOpts.SetMetricsScope(scope))
+	processor := packet.NewProcessor(queue, aggregator, opts.WorkerPoolSize(), clockOpts, instrumentOpts)
+	s := &server{
+		address:      address,
+		opts:         opts,
+		log:          instrumentOpts.Logger(),
+		iteratorPool: opts.IteratorPool(),
+		queue:        queue,
+		processor:    processor,
+		metrics:      newServerMetrics(scope),
+	}
+
+	// Set up the connection functions
+	s.addConnectionFn = s.addConnection
+	s.removeConnectionFn = s.removeConnection
+	s.handleConnectionFn = s.handleConnection
+
+	// Start reporting metrics
+	go s.reportMetrics()
+
+	return s
+}
+
+func (s *server) ListenAndServe() (xclose.SimpleCloser, error) {
+	listener, err := net.Listen("tcp", s.address)
+	if err != nil {
+		return nil, err
+	}
+	s.listener = listener
+	go s.serve()
+	return s, nil
+}
+
+func (s *server) serve() error {
+	connCh, errCh := xnet.StartAcceptLoop(s.listener, s.opts.Retrier())
+	for conn := range connCh {
+		if !s.addConnectionFn(conn) {
+			conn.Close()
+		} else {
+			s.wgConns.Add(1)
+			go s.handleConnectionFn(conn)
+		}
+	}
+	return <-errCh
+}
+
+func (s *server) Close() {
+	s.Lock()
+	if atomic.LoadInt32(&s.closed) == 1 {
+		s.Unlock()
+		return
+	}
+	atomic.StoreInt32(&s.closed, 1)
+	openConns := make([]net.Conn, len(s.conns))
+	copy(openConns, s.conns)
+	s.Unlock()
+
+	// Close all open connections
+	for _, conn := range openConns {
+		conn.Close()
+	}
+
+	// Close the listener
+	if s.listener != nil {
+		s.listener.Close()
+	}
+
+	// Wait for all connection handlers to finish
+	s.wgConns.Wait()
+
+	// Close the packet queue
+	s.queue.Close()
+
+	// Close the packet processor
+	s.processor.Close()
+}
+
+func (s *server) addConnection(conn net.Conn) bool {
+	s.Lock()
+	defer s.Unlock()
+
+	if atomic.LoadInt32(&s.closed) == 1 {
+		return false
+	}
+	s.conns = append(s.conns, conn)
+	atomic.AddInt32(&s.numConns, 1)
+	return true
+}
+
+func (s *server) removeConnection(conn net.Conn) {
+	s.Lock()
+	defer s.Unlock()
+
+	numConns := len(s.conns)
+	for i := 0; i < numConns; i++ {
+		if s.conns[i] == conn {
+			// Move the last connection to i and reduce the number of connections by 1
+			s.conns[i] = s.conns[numConns-1]
+			s.conns = s.conns[:numConns-1]
+			atomic.AddInt32(&s.numConns, -1)
+			return
+		}
+	}
+}
+
+func (s *server) handleConnection(conn net.Conn) {
+	defer func() {
+		conn.Close()
+		s.removeConnectionFn(conn)
+		s.wgConns.Done()
+	}()
+
+	it := s.iteratorPool.Get()
+	it.Reset(bufio.NewReader(conn))
+	defer it.Close()
+
+	// Iterate over the incoming metrics stream and queue up metrics
+	for it.Next() {
+		metric, policies := it.Value()
+		if err := s.queue.Enqueue(packet.Packet{Metric: metric, Policies: policies}); err != nil {
```

Comment:
Can we benchmark this then?  

I don't see why we should funnel data from thousnads of asynchronous driven network IO goroutines into a single contended channel before coming out again on the otherside for not much reason other than for code isolation/structure.